pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1874,15 +1870,21 @@ func buildListener(opts buildListenerOpts) *xdsapi.Listener {
 	listener := &xdsapi.Listener{
 		// TODO: need to sanitize the opts.bind if its a UDS socket, as it could have colons, that envoy
 		// doesn't like
-		Name:            fmt.Sprintf("%s_%d", opts.bind, opts.port),
-		Address:         util.BuildAddress(opts.bind, uint32(opts.port)),
-		ListenerFilters: listenerFilters,
-		FilterChains:    filterChains,
-		DeprecatedV1:    deprecatedV1,
+		Name:             fmt.Sprintf("%s_%d", opts.bind, opts.port),
+		Address:          util.BuildAddress(opts.bind, uint32(opts.port)),
+		ListenerFilters:  listenerFilters,
+		FilterChains:     filterChains,
+		DeprecatedV1:     deprecatedV1,
+		TrafficDirection: trafficDirection,
 	}
 
-	if util.IsIstioVersionGE13(opts.proxy) {
-		listener.ListenerFiltersTimeout = gogo.DurationToProtoDuration(opts.env.Mesh.ProtocolDetectionTimeout)
+	if util.IsIstioVersionGE13(opts.proxy) && !isGateway {
+		if trafficDirection == core.TrafficDirection_OUTBOUND {
+			listener.ListenerFiltersTimeout = gogo.DurationToProtoDuration(opts.env.Mesh.ProtocolDetectionTimeout)
+		} else {
+			listener.ListenerFiltersTimeout = ptypes.DurationProto(features.InboundProtocolDetectionTimeout)
```

Comment:
why do you need to separate the two?

Reply:
I dont think we should separate the two. The idea of introducing that mesh config option was so that people can control. Switching to env vars goes back to non-configurable mode. As far as the current problem is concerned, simply disabling the timeout for gateways is sufficient. From that point of view, can't you simply look at the listener type and decide whether to enable the timeout or not?

Reply:
We would like to provide two variables for timeout. One for inbound and another for outbound.

Reply:
yes but why? what is the benefit of setting inbound timeout on a sidecar to 50s and breaking all mysql traffic? and why make one as env var while the other is mesh config - misleading. And how is this related to the problem solved by the PR?

Reply:
+1 to what @rshriram said wrt different types and config inputs.

It would be good idea to have different values, since the distance between client and outbound sidecar is quite different between client and inbound sidecar, but this can be done in a separate PR.